### PR TITLE
feat(protocol-designer): add dispense air gap to distribute command creator

### LINKED ISSUE #6510: PD: Generate commands for air gap dispense (multi dispense)
# Overview

Add air gap dispense atomic commands for multi dispense path

# Implementation details

See #6711 for step order

Note: This should be thoroughly and rigorously unit tested

# Design

See the flowchart below for more information:

https://www.figma.com/file/0dTEbGNsKYgNwKEW6LLfzJ/Airgap-Delay?node-id=453%3A6016

### PR DESCRIPTION
**Pull request by shlokamin**
# Overview
This PR adds dispense air gap commands to the distribute compound command creator. 

closes #6510 

# Changelog
- Adds dispense air gap commands to the distribute compound command creator. 

# Review requests
- Code review
- Try to follow the tests, I'm pretty sure I covered every possibility but take your time reading through

# Risk assessment

Medium, noodles around with a compound command creator (responsible for actually generating commands to be fed into the robot)

### CODE REVIEW

File: protocol-designer/fixtures/protocol/5/doItAllV3.json [hunk 1]
```diff
@@ -2682,16 +2682,6 @@
...
-      "command": "blowout",
-      "params": {
-        "pipette": "0b3f2210-75c7-11ea-b42f-4b64e50f43e5",
-        "labware": "trashId",
-        "well": "A1",
-        "flowRate": 46.43,
-        "offsetFromBottomMm": 0
-      }
-    },
     {
```
Comment (shlokamin) [inline protocol-designer/fixtures/protocol/5/doItAllV3.json]: removed this because this blowout is unnecessary since we will be dropping tip in trash anyways

File: protocol-designer/src/step-generation/commandCreators/compound/distribute.js [hunk 2]
```diff
@@ -243,26 +276,61 @@ export const distribute: CommandCreator<DistributeArgs> = (
...
-        ]
+      let moreWellsRemaining = false
+      if (destWellChunk.length < maxWellsPerChunk) {
+        moreWellsRemaining = false
+      } else if (
+        args.destWells.length -
+        destWellChunk.length * (chunkIndex + 1)
+      ) {
+        moreWellsRemaining = true
+      }
```
Comment (shlokamin) [inline protocol-designer/src/step-generation/commandCreators/compound/distribute.js]: I don't love this, but I couldn't think of another way around this. I need to somehow know whether or not I have more wells remaining so that I know I can do an air gap after dispense. I.e. to determine whether or not this is after the "last transfer" in @howisthisnamenottakenyet's diagram

Comment (shlokamin) [reply protocol-designer/src/step-generation/commandCreators/compound/distribute.js]: @IanLondon @Kadee80 if you know of a cleaner way to do this lmk because I'd love to nuke this code. I don't like the cognitive overhead it requires.

File: protocol-designer/src/step-generation/commandCreators/compound/distribute.js [hunk 3]
```diff
@@ -243,26 +276,61 @@ export const distribute: CommandCreator<DistributeArgs> = (
...
+
+      let dispenseAirGapLabware = args.destLabware
+      let dispenseAirGapWell = last(destWellChunk)
+
+      if (blowoutLocation === SOURCE_WELL_BLOWOUT_DESTINATION) {
+        dispenseAirGapLabware = args.sourceLabware
+        dispenseAirGapWell = args.sourceWell
```
Comment (shlokamin) [inline protocol-designer/src/step-generation/commandCreators/compound/distribute.js]: I'll extract this with a declarative name, this is hard to follow imo

File: protocol-designer/src/step-generation/__tests__/transfer.test.js [hunk 4]
```diff
@@ -2822,8 +2822,8 @@ describe('advanced options', () => {
           command: 'airGap',
           params: {
             pipette: 'p300SingleId',
-            labware: 'destPlateId',
-            well: 'B1',
+            labware: 'sourcePlateId',
+            well: 'A1',
             volume: 3,
```
Comment (shlokamin) [inline protocol-designer/src/step-generation/__tests__/transfer.test.js:2836]: This and the snippet above are changed because when blowing out from a source well, we must also air gap from the source too

File: protocol-designer/src/step-generation/commandCreators/compound/distribute.js (hunk 2 above)
Comment (shlokamin) [reply protocol-designer/src/step-generation/commandCreators/compound/distribute.js]: saw @IanLondon had a better way to do this in his PR so I copied it :)

File: protocol-designer/src/step-generation/__tests__/distribute.test.js [hunk 5]
```diff
@@ -1,3 +1,4 @@
+/* eslint-disable jest/no-disabled-tests */
```
Comment (IanLondon) [inline protocol-designer/src/step-generation/__tests__/distribute.test.js]: is this `eslint-disable` supposed to be committed?

File: protocol-designer/src/step-generation/__tests__/distribute.test.js [hunk 6]
```diff
@@ -752,209 +801,652 @@ describe('advanced settings: volume, mix, pre-wet tip, tip touch, tip position',
...
-            flowRate: 2.1,
-          },
-        },
-        {
-          command: 'delay',
-          params: {
-            wait: 11,
-          },
-        },
+        airGapHelper('A1', 31),
```
Comment (IanLondon) [inline protocol-designer/src/step-generation/__tests__/distribute.test.js:816]: nitpick: for comments like `// air gap`, I like to differentiate them as `// aspirate > air gap` now that there's room for ambiguity

Comment (IanLondon) [reply protocol-designer/src/step-generation/__tests__/distribute.test.js:816]: likewise I'd prefer `air gap so no liquid drops off the tip as pipette moves from` to have "dispense > air gap" in there somehow just to make searching / scanning easier. Not a super strong opinion though

File: protocol-designer/src/step-generation/utils/misc.js [hunk 7]
```diff
@@ -277,6 +277,29 @@ export function createTipLiquidState<T>(
   )
 }
 
+// always return destination unless the blowout location is the source
+export const getDispenseAirGapLocation = (args: {|
```
Comment (IanLondon) [inline protocol-designer/src/step-generation/utils/misc.js:281]: is it worth writing tests for this fn, just in case we need to change it later on for some reason?

**Review by IanLondon — APPROVED** [review]
looks good! phew

File: protocol-designer/src/step-generation/__tests__/distribute.test.js (hunk 5 above)
Comment (shlokamin) [reply protocol-designer/src/step-generation/__tests__/distribute.test.js]: no good catch this was an accident!

File: protocol-designer/src/step-generation/__tests__/distribute.test.js (hunk 6 above)
Comment (shlokamin) [reply protocol-designer/src/step-generation/__tests__/distribute.test.js:816]: I like that convention! Will update the comments

File: protocol-designer/src/step-generation/utils/misc.js (hunk 7 above)
Comment (shlokamin) [reply protocol-designer/src/step-generation/utils/misc.js:281]: YES good call!
